File: src/main/java/org/nordix/simplepki/application/port/in/Pki.java
/*-
 * ============LICENSE_START=======================================================
 *  Copyright (C) 2022 Nordix Foundation.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * ============LICENSE_END=========================================================
 */
package org.nordix.simplepki.application.port.in;

import org.bouncycastle.pkcs.PKCS10CertificationRequest;

import java.io.IOException;
import java.security.cert.CertificateException;
import java.security.cert.X509CRL;
import java.security.cert.X509Certificate;
import java.util.Date;
import java.util.Optional;
import java.util.function.LongPredicate;

public interface Pki {
    X509Certificate getCaCert();

    Pki.CrlBuilder crlBuilder();

    X509Certificate sign(PKCS10CertificationRequest csr)
        throws Exception;

    X509Certificate getCertificate(String serialNumber)
        throws CertificateException, IOException;

    X509Certificate getCertificate(long serialNumber)
        throws CertificateException, IOException;

    boolean revoke(String serialNumber, Date date);

    boolean revoke(long serialNumber, Date date);

    interface CrlBuilder {
        CrlBuilder filterByUpdateTime(LongPredicate predicate);

        Optional<X509CRL> build()
            throws Exception;
    }
}

File: src/main/kotlin/org/nordix/simplepki/application/port/in/Pki.kt
/*-
 * ============LICENSE_START=======================================================
 *  Copyright (C) 2022 Nordix Foundation.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * ============LICENSE_END=========================================================
 */
package org.nordix.simplepki.application.port.`in`

import org.bouncycastle.pkcs.PKCS10CertificationRequest
import java.security.cert.X509CRL
import java.security.cert.X509Certificate
import java.util.*

interface Pki {

    val caCert: X509Certificate

    fun crlBuilder(): CrlBuilder

    fun sign(csr: PKCS10CertificationRequest): X509Certificate

    fun getCertificate(serialNumber: String): X509Certificate

    fun getCertificate(serialNumber: Long): X509Certificate

    fun revoke(serialNumber: String, date: Date): Boolean

    fun revoke(serialNumber: Long, date: Date): Boolean

    interface CrlBuilder {
        fun editionTime(): Long

        fun build(): X509CRL
    }
}
